### ISSUE TITLE
A document on How Things Work

### ISSUE THREAD

**Issue by Sumith1896** [sympy/symengine#443]
We currently don't have good documentation, the [design](https://github.com/sympy/symengine/blob/master/doc/design.md) doc is helpful but not exhaustive considering some recent developments.
When I was going over the code, there were instances where I couldn't figure things and had to go to varied sources. For example: some info in visitor pattern etc.
The plan is to add a markdown file maybe, where all of us put in our inputs and though some of details found in this can be found reading some source elsewhere, having such a document attached to SymEngine is good.

**Comment by Sumith1896** [comment]
A really rough base structure has been framed here https://github.com/sympy/symengine/pull/444.

**Comment by certik** [comment]
Yes, that would be good. We can start with a markdown document for now. Later we should have probably Sphinx documentation and have everything there.

**Comment by Sumith1896** [comment]
@certik When this expands, we'll convert it Sphinx, and also documentation will happen as we do in SymPy.
Till then let's update the markdown file, and every PR with major change is accompanied with the documentation update.

**Comment by certik** [comment]
I agree.
